book/src/dev/rfcs/XXXX-asynchronous-script-verification.md

```diff
@@ -0,0 +1,231 @@
+worker task and passing messages to it over an multi-producer single-consumer
+(mpsc) channel. The worker task recieves messages and makes `Service::call`s.
+The `Service::call` method returns a `Future`, and the service is allowed to
+decide how much work it wants to do synchronously (in `call`) and how much
+work it wants to do asynchronously (in the `Future` it returns).
+
+This means that our external API ensures that the state service sees a
+linearized sequence of state requests, although the exact ordering is
+unpredictable when there are multiple senders making requests.
+
+In the second category, the Sled API presents itself synchronously, but
+database and tree handles are clonable and can be moved between threads. All
+that's required to process some request asynchronously is to clone the
+appropriate handle, move it into an async block, and make the call as part of
+the future. (We might want to use Tokio's blocking API for this, but that's a
+side detail).
+
+Because the state service has exclusive access to the sled database, and the
+state service sees a linearized sequence of state requests, we have an easy
+way to opt in to asynchronous database access. We can perform sled operations
+synchronously in the `Service::call`, waiting for them to complete, and be
+sure that all future requests will see the resulting sled state. Or, we can
+perform sled operations asynchronously in the future returned by
+`Service::call`.
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+Now, returning to the UTXO lookup problem, we can map out the possible states
+with this restriction in mind. This description assumes that UTXO storage is
+split into disjoint sets, one in-memory (e.g., blocks after the reorg limit)
+and the other in sled (e.g., blocks after the reorg limit). The details of
+this storage are not important for this design, only that the two sets are
+disjoint.
+
+When the state service processes a `Request::AwaitUtxo(OutPoint)` referencing
+some UTXO `u`, there are three disjoint possibilities:
+
+1. `u` is already contained in an in-memory block storage;
+2. `u` is already contained in the sled UTXO set;
+3. `u` is not yet known to the state service.
+
+In case 3, we need to queue `u` and scan all *future* blocks to see whether
+they contain `u`. However, if we have a mechanism to queue `u`, we can
+perform check 2 asynchronously, because restricting to synchronous writes
+means that any async read will return the current or later state. If `u` was
+in the sled UTXO set when the request was processed, the only way that an
+async read would not return `u` is if the UTXO were spent, in which case the
+service is not required to return a response.
```

Comment:
Question: What happens when we move UTXOs from in-memory storage to sled storage?

Do we need to hold a lock over both states while we modify them?
Or is it sufficient to ensure that sled has any additional UTXOs, before we remove those UTXOs from in-memory storage?

I'd like to see this case spelled out in a bit more detail, because it's not obvious to me from the text.

Reply:
I don't think that this can happen, because that movement can't happen simultaneously with checking any in-memory data. It would be a subsequent service call, after any in-memory checks had happened.